mmlearn/datasets/processors/masking.py

```diff
@@ -264,3 +265,134 @@ def apply_masks(
+        self.height = self.input_size[0] // self.patch_size
+        self.width = self.input_size[1] // self.patch_size
+
+    def _sample_block_size(
+        self,
+        generator: torch.Generator,
+        scale: Tuple[float, float],
+        aspect_ratio: Tuple[float, float],
+    ) -> Tuple[int, int]:
+        """Sample the size of the mask block based on scale and aspect ratio."""
+        _rand = torch.rand(1, generator=generator).item()
+        min_s, max_s = scale
+        mask_scale = min_s + _rand * (max_s - min_s)
+        max_keep = int(self.height * self.width * mask_scale)
+
+        min_ar, max_ar = aspect_ratio
+        aspect_ratio_val = min_ar + _rand * (max_ar - min_ar)
+
+        h = int(round(math.sqrt(max_keep * aspect_ratio_val)))
+        w = int(round(math.sqrt(max_keep / aspect_ratio_val)))
+
+        h = min(h, self.height - 1)
+        w = min(w, self.width - 1)
+
+        return h, w
+
+    def _sample_block_mask(
+        self,
+        b_size: Tuple[int, int],
+        acceptable_regions: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Sample a mask block."""
+        h, w = b_size
+        top = torch.randint(0, self.height - h, (1,)).item()
+        left = torch.randint(0, self.width - w, (1,)).item()
+        mask = torch.zeros((self.height, self.width), dtype=torch.int32)
+        mask[top : top + h, left : left + w] = 1
+
+        mask_complement = torch.ones((self.height, self.width), dtype=torch.int32)
+        mask_complement[top : top + h, left : left + w] = 0
+
+        return mask.flatten(), mask_complement.flatten()
+
+    def __call__(
+        self,
+        example: torch.Tensor,
+    ) -> Dict[str, Any]:
+        """Generate encoder and predictor masks for a single example.
+
+        Parameters
+        ----------
+        example : torch.Tensor
+            The input data example (e.g., image tensor) to process.
+
+        Returns
+        -------
+        Dict[str, Any]
+            A dictionary of the input example, encoder masks, and predictor masks.
+        """
+        seed = torch.randint(
+            0, 2**32, (1,)
+        ).item()  # Sample random seed for reproducibility
+        g = torch.Generator().manual_seed(seed)
+
+        # Sample block sizes
+        p_size = self._sample_block_size(
+            generator=g, scale=self.pred_mask_scale, aspect_ratio=self.aspect_ratio
+        )
+        e_size = self._sample_block_size(
+            generator=g, scale=self.enc_mask_scale, aspect_ratio=(1.0, 1.0)
+        )
+
+        # Generate predictor masks
+        masks_pred, masks_enc = [], []
+        for _ in range(self.npred):
+            mask_p, _ = self._sample_block_mask(p_size)
+            masks_pred.append(mask_p)
+
+        # Generate encoder masks
+        for _ in range(self.nenc):
+            mask_e, _ = self._sample_block_mask(e_size)
+            masks_enc.append(mask_e)
+
+        return {
+            "input": example,
```

Comment:
One last thing: If this class is intended to be used with the deafult data collator, returning `example` under the 'input' key won't work. See [this](https://github.com/VectorInstitute/mmlearn/blob/main/mmlearn/datasets/core/data_collator.py#L54) for why.

You can either change the implementation of this class or change the implementation in the collator. That's if this class is intended to be used with the collator.